plugins/scale/wayfire/plugins/scale-transform.hpp

```diff
@@ -0,0 +1,181 @@
+        view->damage();
+    }
+
+    void add_overlay(std::shared_ptr<scale_view_overlay>&& ol, int z_order,
+        const padding& pad = {0, 0, 0, 0})
+    {
+        add_overlay(ol, z_order, pad);
+    }
+
+    /* remove an existing overlay */
+    void rem_overlay(scale_view_overlay *ol)
+    {
+        view->damage();
+        overlays.remove_if([ol] (const auto& other)
+        {
+            return other->ol.get() == ol;
+        });
+
+        /* recalculate padding */
+        pad = {0, 0, 0, 0};
+        for (const auto& ol : overlays)
+        {
+            extend_padding(pad, ol->pad);
+        }
+
+        view->damage();
+    }
+
+    /* get the view being transformed (it is protected in view_2D) */
+    wayfire_view get_transformed_view() const
+    {
+        return view;
+    }
+
+    /**
+     * Transformer name used by scale. This can be used by other plugins to find
+     * scale's transformer on a view.
+     */
+    static std::string transformer_name()
+    {
+        return "scale";
+    }
+
+    /**
+     * Transform a box, including the current transform, but not the padding.
+     */
+    wlr_box trasform_box_without_padding(wlr_box box)
+    {
+        box = view->transform_region(box, this);
+        wlr_box view_box = view->get_bounding_box(this);
+        return view_transformer_t::get_bounding_box(view_box, box);
+    }
+
+    /**
+     * Transform a region and add padding to it.
+     * Note: this will pad any transformed region, not only if it corresponds to
+     * the view's bounding box.
+     */
+    wlr_box get_bounding_box(wf::geometry_t view, wlr_box region) override
+    {
+        region    = view_transformer_t::get_bounding_box(view, region);
+        region.x -= pad.left;
+        region.y -= pad.top;
+        region.width  += pad.left + pad.right;
+        region.height += pad.top + pad.bottom;
+        return region;
+    }
+
+    const padding& get_padding() const
+    {
+        return pad;
+    }
+
+    /**
+     * extend padding in target to ensure it is at least as large as other
+     */
+    static void extend_padding(padding& target, const padding& other)
+    {
+        target.top    = std::max(target.top, other.top);
+        target.left   = std::max(target.left, other.left);
+        target.bottom = std::max(target.bottom, other.bottom);
+        target.right  = std::max(target.right, other.right);
+    }
+
+  protected:
+    /* extra info stored together with the overlay callbacks */
+    struct overlay
+    {
+        std::shared_ptr<scale_view_overlay> ol; /* callback function */
+        int z_order; /* order in which overalys are applied */
+        padding pad; /* extra padding needed by this overlay */
+        overlay(std::shared_ptr<scale_view_overlay>& ol_, int z_order_,
+            const padding& pad_) :
+            ol(ol_), z_order(z_order_), pad(pad_)
+        {}
+    };
+
+    std::list<std::unique_ptr<overlay>> overlays; /* list of active overlays */
```

Comment:
Seems to me that overlays are movable and copyable - do we need to keep them in `unique_ptr`s ?

Reply:
I was unsure and I switched to using shared_ptr everywhere. Now I feel it should be unique_ptr, since one overlay will go together with one view transformer. If the caller wants to share state / resources between overlays, they can do this by inheriting from the overlay struct and adding shared_ptr members.
Also, I'm not sure if the callbacks should be std::functions, or it might be better to use virtual functions (if users of overlays will typically inherit anyway).

Reply:
I see, I have missed this reorganization - I was reviewing commit-by-commit, so this particular comment is outdated.

In any case, shared_ptr is not semantically correct here. It implies shared ownership, however, an overlay should never remain in the transformer if the entity which added it is destroyed. So a unique_ptr/observer_ptr combo sounds better.
Plus, if you push observer-ptrs, you can make several such pointers from the same unique_ptr, thus sharing overlays between transformers, if you wanted to.

Another note: please rename `overlay` and `padding` with `_t` versions.

Reply:
I'm leaning toward `unique_ptr` -- this would allow to create, add and "forget" an overlay (and anyway, in most cases, there needs to be some extra data that is unique to every view).

Reply:
Yeah, this is what we do for ex. for view transformers and custom data. There are some pitfalls though - you still need to ensure all of these are removed when the plugin is unloading, otherwise, it could happen that the .so file is unloaded (and thus the destructor cannot run anymore).

Reply:
That's true. For now, the title-overlay is compiled in the same library as scale, so this will not be a problem, but for other uses, this means there is a need to keep track of overlays added (but I'm not sure if using `observer_ptr` or `shared_ptr` would solve it).